### ISSUE TITLE
Template with result_of only works in same cell in which it's defined

### ISSUE THREAD

**Issue by stevenrbrandt** [root-project/root#7952]
This cell evaluates
```
#include <iostream>
#include <type_traits>
#include <functional>
template<typename T>
T fun2(std::function<T()> f) {
    T t = f();
    return t;
}
template<>
void fun2(std::function<void()> f) {
    f();
}
template<typename F>
typename std::result_of<F()>::type fun(F f) {
    typedef typename std::result_of<F()>::type ftype;
    return fun2<ftype>(f);
}

// This call succeeds
auto a = fun([](){ return 42; });
```
However, the following in a separate cell fails
```
// This call kills the kernel
auto b = fun([](){ return 42; });
```
Note that I am using the cling_2020-11-05_ROOT-fedora32 in a fedora32 docker image.

**Comment by guitargeek** [comment]
Hi @stevenrbrandt, sorry for coming back to this so late!

Is this issue still relevant anymore? It could be gone in ROOT 6.30, where the minimum C++ standard is C++17, or maybe it's gone at least in ROOT `master` with the upgrade to LLVM 16.

Can you still reproduce this issue with the newest ROOT?

**Comment by dpiparo** [comment]
Yes, this is fixed in 6.30 - just tested on SWAN.
